cpp/CMakeLists.txt

```diff
@@ -37,300 +79,233 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
-endif (CUDF_INCLUDE AND CUDF_LIBRARY)
+# * find CUDAToolkit package
+# * determine GPU architectures
+# * enable the CMake CUDA language
+# * set other CUDA compilation flags
+include(cmake/Modules/ConfigureCUDA.cmake)
 
 ###################################################################################################
-# - gdal ------------------------------------------------------------------------------------------
+# - dependencies ----------------------------------------------------------------------------------
 
+# find gdal
 find_package(GDAL REQUIRED)
-
-if(NOT GDAL_FOUND)
-    message(FATAL_ERROR "GDAL not found, please check your settings.")
-endif(NOT GDAL_FOUND)
+# add third party dependencies using CPM
+include(cmake/thirdparty/CUSPATIAL_GetCPM.cmake)
+# find or add RMM
+include(cmake/thirdparty/CUSPATIAL_GetRMM.cmake)
+# find or add GTest
+include(cmake/thirdparty/CUSPATIAL_GetGTest.cmake)
+# find or add cuDF
+include(cmake/thirdparty/CUSPATIAL_GetCUDF.cmake)
 
 ###################################################################################################
-# - add gtest -------------------------------------------------------------------------------------
+# - library targets -------------------------------------------------------------------------------
 
-if(BUILD_TESTS)
-    include(CTest)
-    include(ConfigureGoogleTest)
+add_library(cuspatial SHARED
+    src/indexing/construction/point_quadtree.cu
+    src/interpolate/cubic_spline.cu
+    src/io/shp/polygon_shapefile_reader.cpp
+    src/io/shp/polygon_shapefile_reader.cu
+    src/join/quadtree_point_in_polygon.cu
+    src/join/quadtree_point_to_nearest_polyline.cu
+    src/join/quadtree_poly_filtering.cu
+    src/spatial/polygon_bounding_box.cu
+    src/spatial/polyline_bounding_box.cu
+    src/spatial/point_in_polygon.cu
+    src/spatial_window/spatial_window.cu
+    src/spatial/haversine.cu
+    src/spatial/hausdorff.cu
+    src/spatial/lonlat_to_cartesian.cu
+    src/trajectory/derive_trajectories.cu
+    src/trajectory/trajectory_bounding_boxes.cu
+    src/trajectory/trajectory_distances_and_speeds.cu
+)
+
+set_target_properties(cuspatial
+    PROPERTIES BUILD_RPATH                         "\$ORIGIN"
+               # set target compile options
+               CXX_STANDARD                        14
+               CXX_STANDARD_REQUIRED               ON
+               CUDA_STANDARD                       14
+               CUDA_STANDARD_REQUIRED              ON
+               POSITION_INDEPENDENT_CODE           ON
+               INTERFACE_POSITION_INDEPENDENT_CODE ON
+)
+
+target_compile_options(cuspatial
+            PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${CUSPATIAL_CXX_FLAGS}>"
+                    "$<$<COMPILE_LANGUAGE:CUDA>:${CUSPATIAL_CUDA_FLAGS}>"
+)
+
+target_compile_definitions(cuspatial
+            PUBLIC "$<$<COMPILE_LANGUAGE:CXX>:${CUSPATIAL_CXX_DEFINITIONS}>"
+                   "$<$<COMPILE_LANGUAGE:CUDA>:${CUSPATIAL_CUDA_DEFINITIONS}>"
```

Comment:
Do we need to guard this with a `BUILD_INTERFACE` generator expression to prevent leaking these in the case of someone building on top of libcuspatial with non-cuda things?